plugins/llvm/llvm_binary.hpp

```diff
@@ -235,75 +251,87 @@ OutputIterator read(symbol_iterator begin,
+    read(obj->symbols().begin(),
+         obj->symbols().end(),
          std::back_inserter(symbols));
+	*/
     return symbols;
 }
 
-
-
-
 std::vector<symbol> read(const COFFObjectFile * obj) {
-    std::vector<symbol> symbols;
-
-    const pe32_header *pe32;
-    if (error_code err = obj->getPE32Header(pe32))
-        llvm_binary_fail(err);
-
-    for (auto it = obj->begin_symbols(); it != obj->end_symbols(); ++it) {
-        auto sym = obj->getCOFFSymbol(it);
-
-        if (!sym) llvm_binary_fail("not a coff symbol");
-
-        const coff_section *sec = nullptr;
-        if (sym->SectionNumber == COFF::IMAGE_SYM_UNDEFINED)
-            continue;
-
-        if (error_code ec = obj->getSection(sym->SectionNumber, sec))
-            llvm_binary_fail(ec);
-
-        if (!sec) continue;
-
-        uint64_t size = (sec->VirtualAddress + sec->SizeOfRawData) - sym->Value;
-
-        for (auto it = obj->begin_symbols(); it != obj->end_symbols(); ++it) {
-            auto next = obj->getCOFFSymbol(it);
-            if (next->SectionNumber == sym->SectionNumber) {
-                auto new_size = next->Value > sym->Value ?
-                    next->Value - sym->Value : size;
-                size = new_size < size ? new_size : size;
-            }
-        }
-
-        auto addr = sec->VirtualAddress + pe32->ImageBase + sym->Value;
-        symbols.push_back(symbol(*it,addr,size));
-    }
-    return symbols;
+  	std::vector<symbol> symbols;
+	auto symbol_sizes = llvm::object::computeSymbolSizes(*obj);
+
+	const pe32_header *pe32;
+	if (std::error_code err = obj->getPE32Header(pe32))
+		llvm_binary_fail(err);
+
+	ObjectFile::symbol_iterator_range symbol_range = obj->symbols();
+	for (auto it : symbol_range) {
+		auto sym = obj->getCOFFSymbol(it);
+
+		//if (!sym) ...
+		
+		const coff_section *sec = nullptr;
+		if (sym.getSectionNumber() == COFF::IMAGE_SYM_UNDEFINED)
+			continue;
+
+		if (std::error_code ec = obj->getSection(sym.getSectionNumber(), sec))
+			llvm_binary_fail(ec);
+
+		if (!sec) continue;
+
+	  	uint64_t size = (sec->VirtualAddress + sec->SizeOfRawData) - sym.getValue();
+
+   		for (auto it : symbol_range) {
+	  		auto next = obj->getCOFFSymbol(it);
+   			if (next.getSectionNumber() == sym.getSectionNumber()) {
+   				auto new_size = next.getValue() > sym.getValue() ?
+   					next.getValue() - sym.getValue() : size;
+   				size = new_size < size ? new_size : size;
+   			}
+   		}
+
+		auto addr = sec->VirtualAddress + pe32->ImageBase + sym.getValue();
+		symbols.push_back(symbol(it,addr,size));
+   	}
+	return symbols;
 }
 
 template <typename ELFT>
 std::vector<symbol> read(const ELFObjectFile<ELFT>* obj) {
-    int size1 = utils::distance(obj->begin_symbols(),
-                                obj->end_symbols());
-    int size2 = utils::distance(obj->begin_dynamic_symbols(),
-                                obj->end_dynamic_symbols());
+    //int size1 = std::distance(obj->symbols().begin(), obj->symbols().end());
```

Comment:
The body of this function is essentially the same as the body of a generic version.